### PR TITLE
Ensures CJS/ESM bundle is compatible with Alpha (sonic), pagesJS (vite), and CRA (webpack)

### PR DESCRIPTION
**Pull request by yen-tt**
**Investigation:**
This PR attempts to fix the following error in alpha/sonic:
Screenshot 2023-08-31 at 4 40 11 PM
This was a result of adding `"type": "module"` to our esm bundle for our library to be compatible with vite/pagesJS.
Vite, and bundlers such as webpack, follow node's module resolution algorithm. and NodeJS have certain rules on how it recognize files as ES modules (https://nodejs.org/api/packages.html#determining-module-system) -- which include adding `"type": "module"` as an option.

The fix is to update all of our paths to be explicit in order to include the mjs/js extensions, including  (e.g. `./components` to `./components/index.js` and `../icons/DualSync` to `../icons/DualSync.js`).

This currently can't be done by typescript compiler (tsc) as it's strictly a nodejs behavior and they don't want to support that ([long issue thread here](https://github.com/microsoft/TypeScript/issues/18442#issuecomment-408654005)), which is frustrating. So we would have to do it manually or have a script to update import/export paths in the final bundle generated by tsc. This is not ideal and can be error prone.

**Solution:**
I decided to **replace tsc with rollup to bundle our library and append .mjs extension to our final esm bundle**. Using `mjs` extension instead of `.js`  also remove the need to do `"type": "module"` in our esm package.json, which previously introduce unnecessary caveats and one-off script to inject it into our esm bundle.

NOTE: the default **interop** behavior for rollup is not compatible with alpha/sonic. As such, the rollup config in this PR uses `auto` to follow Typescript's `esModuleInterop` behavior in how it transpile named, default, and dynamic imports of external deps (the issue with alpha is related to how react-textarea-autosize was imported into ChatInput)

NOTE: updated `tsconfig.json` to use `"jsx": "react"` instead of `"jsx": "react-jsx"` because of the way jsx-runtime is backported to react 16/17 without explicit exports field ([React github issue here](https://github.com/facebook/react/issues/20235#issuecomment-1095345193)). We would either need to export two separate bundles for latest and legacy versions in order to continue outputting the JSX syntactic sugar OR we could just output `React.createElement` directly. This is common for many React libraries that support older versions like React 16. (e.g. [ant design](https://github.com/ant-design/ant-design/blob/master/tsconfig.json#L15), [blueprint UI](https://github.com/palantir/blueprint/blob/develop/config/tsconfig.base.json#L9), [evergreen UI](https://github.com/segmentio/evergreen/blob/master/tsconfig.json#L6))

J=CLIP-520
TEST=manual

see that the new build works with:
- the local test-site of the component lib repo 
- pagejs/vite (released v0.6.0-alpha.38.6 to install and test)
- sonic/alpha (released v0.6.0-alpha.38.6 to install and test)

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -20805,6 +20856,85 @@
...
+      },
+      "bin": {
+        "semver": "bin/semver.js"
+      },
+      "engines": {
+        "node": ">=10"
+      }
+    },
+    "node_modules/rollup-plugin-typescript2/node_modules/tslib": {
+      "version": "2.6.2",
```
[the remark this answers is not in the document]
Comment (yen-tt) [reply package-lock.json:20918]: this is resolved. Legal approved of the use case of tslib within the build process with license 0BSD.

**Review by nbramblett — APPROVED** [review]

File: package-lock.json [hunk 2]
```diff
@@ -19358,6 +21461,31 @@
       "integrity": "sha512-LDJzPVEEEPR+y48z93A0Ed0yXb8pAByGWo/k5YYdYgpY2/2EsOsksJrq7lOHxryrVOn1ejG6oAp8ahvOIQD8sw==",
       "dev": true
     },
+    "node_modules/path-scurry": {
+      "version": "1.10.1",
```
[the remark this answers is not in the document]
Comment (yen-tt) [reply package-lock.json]: reverted -- this no longer applied.

File: package-lock.json [hunk 3]
```diff
@@ -15483,6 +17568,24 @@
         "node": ">=8"
       }
     },
+    "node_modules/jackspeak": {
+      "version": "2.3.3",
```
[the remark this answers is not in the document]
Comment (yen-tt) [reply package-lock.json]: reverted -- this no longer applied.

File: package-lock.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (yen-tt) [reply package-lock.json]: reverted -- this no longer applied.

**Review by nbramblett — APPROVED** [review]

**Pull request merged**